src/ResourceManager/Reservations/Commands.Reservations/Common/AzureReservationsAutoRegisterCmdletBase.cs

```diff
@@ -0,0 +1,51 @@
+﻿using System;
+using System.Collections.Generic;
+using System.Linq;
+using System.Text;
+using System.Threading.Tasks;
+
+using Microsoft.Azure.Commands.Common.Authentication;
+using Microsoft.Azure.Commands.Common.Authentication.Models;
+using Microsoft.Azure.Commands.ResourceManager.Common;
+using Microsoft.Azure.Management.Reservations;
+using Microsoft.Azure.Commands.Common.Authentication.Abstractions;
+using Microsoft.Azure.Management.Internal.Resources;
+
+namespace Microsoft.Azure.Commands.Reservations.Common
+{
+    /// <summary>
+    /// Base class of Azure Reservations Auto Register Cmdlet.
+    /// </summary>
+    public abstract class AzureReservationsAutoRegisterCmdletBase : AzureReservationsCmdletBase
+    {
+        //Register for compute is required for Reservations commands inheriting this class
+        private string ComputeProviderNamespace = "Microsoft.Compute";
+
+        protected abstract bool ShouldRegister { get; }
+
+        protected override void BeginProcessing()
+        {
+            AzureSession.Instance.ClientFactory.RemoveHandler(typeof(AzureReservationsAutoRegisterDelegatingHandler));
+            IAzureContext context;
+            if (TryGetDefaultContext(out context)
+                && context.Account != null
+                && context.Subscription != null)
+            {
+                AzureSession.Instance.ClientFactory.AddHandler(new AzureReservationsAutoRegisterDelegatingHandler(
```

Comment:
This is unnecessary.  The existing auto-registration infrastructure auto-registers on any creation request that faile due to an unregistered subscription

Reply:
@markcowl the reason we are doing this is because we want to register the subscription with Microsoft.Compute - once this PATCH call comes to RI-RP (Reservations), we make a call to other service (CAS MCIO) which require the subscription to be registered with MS.Compute. So the initial call from Powershell => RI-RP can fail and the existing auto-registration infra might not be able to catch unregistered exception (because that exception is happening between RI-RP and CAS) so we need to manually do registration.

Reply:
Do you have some evidence that this doesn't work in the existing infra?  I would prefer updating the existing auto-registration componenent rather than adding a new one.

Reply:
This is not limited to your cmdlets.  This is applying this to every client created from the client factory,  

IMO, this should not be done in a handler, but should simply be a call using the resource manager client before any cmdlets that need to be registered.

Howver, if you feel that you need this to be a DelegatingHandler for some reason, then you need to apply it to your client after creation has taken place.